domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,202 @@
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/secretbackend"
+	backenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/internal/database"
+)
+
+type upsertOperation struct {
+	secretbackend.UpsertSecretBackendParams
+
+	getBackendStmt *sqlair.Statement
+
+	upsertBackendStmt                 *sqlair.Statement
+	upsertRotationStmt                *sqlair.Statement
+	upsertConfigStmt, clearConfigStmt *sqlair.Statement
+}
+
+// Prepare prepares the sqlair statements.
+func (o *upsertOperation) Prepare() (err error) {
+	if o.ID == "" {
+		return fmt.Errorf("backend ID is missing: %w", backenderrors.NotValid)
+	}
+
+	o.getBackendStmt, err = sqlair.Prepare(`
+SELECT 
+    name                  AS &SecretBackend.name,
+    backend_type          AS &SecretBackend.backend_type,
+	token_rotate_interval AS &SecretBackend.token_rotate_interval
+FROM secret_backend
+WHERE uuid = $M.uuid`, SecretBackend{}, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertBackendStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend
+    (uuid, name, backend_type, token_rotate_interval)
+VALUES (
+    $SecretBackend.uuid,
+    $SecretBackend.name,
+    $SecretBackend.backend_type,
+    $SecretBackend.token_rotate_interval
+)
+ON CONFLICT (uuid) DO UPDATE SET
+    name=EXCLUDED.name,
+    token_rotate_interval=EXCLUDED.token_rotate_interval;`, SecretBackend{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertRotationStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_rotation
+    (backend_uuid, next_rotation_time)
+VALUES (
+    $SecretBackendRotation.backend_uuid,
+    $SecretBackendRotation.next_rotation_time
+)
+ON CONFLICT (backend_uuid) DO UPDATE SET next_rotation_time=EXCLUDED.next_rotation_time;`,
+		SecretBackendRotation{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.clearConfigStmt, err = sqlair.Prepare(`
+DELETE FROM secret_backend_config
+WHERE backend_uuid = $M.uuid AND name NOT IN ($S[:]);`,
+		sqlair.M{}, sqlair.S{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	o.upsertConfigStmt, err = sqlair.Prepare(`
+INSERT INTO secret_backend_config
+    (backend_uuid, name, content)
+VALUES (
+    $SecretBackendConfig.backend_uuid,
+    $SecretBackendConfig.name,
+    $SecretBackendConfig.content
+)
+ON CONFLICT (backend_uuid, name) DO UPDATE SET content = EXCLUDED.content`,
+		SecretBackendConfig{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (o *upsertOperation) validate(ctx context.Context, tx *sqlair.TX) error {
+	var existing SecretBackend
+	err := tx.Query(ctx, o.getBackendStmt, sqlair.M{"uuid": o.ID}).Get(&existing)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		// New insert.
+		if o.Name == "" {
+			return fmt.Errorf("backend name is missing: %w", backenderrors.NotValid)
```

Comment:
Maybe like this
`fmt.Errorf(%w: "backend name is missing", backenderrors.NotValid)

Reply:
I don't see a common pattern in the domain package for error format.
I do see we have more "xxx: %w" errors and I personally prefer this more as well.

Reply:
Ok, I changed my mind, "%w: xxx" is better.

Reply:
Yeah, it was suggested because of the way the final error string would end up looking